keanu-project/src/main/java/io/improbable/keanu/util/Py4jByteArrayConverter.java

```diff
@@ -40,4 +41,62 @@
         byte[] byteArray = booleanBuffer.array();
         return byteArray;
     }
+
+    public static double[] toDoubleArray(byte[] byteArray) {
+        int doubleSizeBytes = Double.SIZE / Byte.SIZE;
+
+        double[] doubles = new double[byteArray.length / doubleSizeBytes];
+        for (int i = 0; i < doubles.length; i++) {
+            doubles[i] = ByteBuffer.wrap(byteArray, i * doubleSizeBytes, doubleSizeBytes).order(ByteOrder.LITTLE_ENDIAN).getDouble();
+        }
+        return doubles;
+    }
+
+    public static int[] toIntegerArray(byte[] byteArray) {
+        int intSizeBytes = Integer.SIZE / Byte.SIZE;
+
+        int[] ints = new int[byteArray.length / intSizeBytes];
+        for(int i = 0; i < ints.length; i++) {
+            ints[i] = ByteBuffer.wrap(byteArray, i * intSizeBytes, intSizeBytes).order(ByteOrder.LITTLE_ENDIAN).getInt();
+        }
+        return ints;
+    }
+
+    public static long[] toLongArray(byte[] byteArray) {
+        int longSizeBytes = Long.SIZE / Byte.SIZE;
+
+        long[] longs = new long[byteArray.length / longSizeBytes];
+        for(int i = 0; i < longs.length; i++) {
+            longs[i] = ByteBuffer.wrap(byteArray, i * longSizeBytes, longSizeBytes).order(ByteOrder.LITTLE_ENDIAN).getLong();
+        }
+        return longs;
+    }
+
+    /**
+     * This function is intended to be called from the Python API with the result of np.packbits(...).toBytes().
+     * In Python, the result of this is a byte array where each boolean is represented by a single bit (0 for false, 1 for true).
+     * When put into a bytearray, these bits may need to be padded with up to 7 bits to make a whole number of bytes.
+     * Therefore, the function requires the parameter numberOfBooleansInArray to know which bits are padding bits and which are booleans.
+     * @param byteArray an array of bytes where booleans (in little endian form) are represented by single bits.
+     * @param numberOfBooleansInArray the number of booleans represented by the bits in the bytearray
+     * @return an array of booleans
+     */
+    public static boolean[] toBooleanArray(byte[] byteArray, int numberOfBooleansInArray) {
+        BitSet bits = BitSet.valueOf(byteArray);
+        boolean[] bools = new boolean[numberOfBooleansInArray];
+        int totalNumberOfBits = byteArray.length * 8;
+        for (int i = bits.nextSetBit(0); i != -1 && i < totalNumberOfBits; i = bits.nextSetBit(i+1)) {
+            int position = getBigEndianPosition(i);
+            if (position < numberOfBooleansInArray) {
```

Comment:
Should we throw an exception if this is not the case?

Reply:
No - in the case that the number of booleans does not fit exactly into n bytes (i.e. the number of bools is not a multiple of 8) then this case can occur completely validly. What it means, is that we are parsing a 'padding bit' and that we should ignore it. Because of the endianness, this could occur whilst there are still further valid bits yet to parse.

Reply:
@icantsayrural as you pointed out the above is wrong. Any extra padding bits should be zero and therefore not appear here as a `nextSetBit`. The behaviour will now throw an error if any set bits are found that are longer than the number of booleans to be encoded.
This behaviour is also now tested